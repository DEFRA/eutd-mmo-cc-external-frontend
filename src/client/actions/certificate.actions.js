import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};