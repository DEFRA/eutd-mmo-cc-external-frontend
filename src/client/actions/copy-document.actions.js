import { showFullPageError, apiCallFailed, isError, CLEAR_ERRORS } from './index';
export const CONFIRM_COPY_DOCUMENT            = 'confirm_copy_document';
export const UNAUTHORISED_COPY_DOCUMENT       = 'unauthorised_copy_document';
export const CLEAR_COPY_DOCUMENT              = 'clear_copy_document';
export const REMOVE_CONFIRM_COPY_CERTIFICATE  = 'remove_confirm_copy_certificate';
export const ADD_ERROR_COPY_CERTIFICATE       = 'add_error_copy_certificate';
export const REMOVE_ERROR_COPY_CERTIFICATE    = 'remove_error_copy_certificate';

export const submitCopyCertificate = (payload) => async (dispatch, _getState, {orchestrationApi}) => {
  const {documentNumber, copyDocumentAcknowledged, voidOriginal, excludeLandings, journey} = payload;
  const config = {headers: {documentnumber: documentNumber}};

  try {
    let ipAddress;
    if (voidOriginal) {
      const ip = await orchestrationApi.get('/client-ip');
      ipAddress = ip.data;
    }

    return await orchestrationApi.post('/confirm-copy-certificate', { copyDocumentAcknowledged, voidOriginal, excludeLandings, journey, ipAddress }, config)
    .then(res => {
      dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: {
        documentNumber: res.data.documentNumber,
        copyDocumentAcknowledged: res.data.copyDocumentAcknowledged,
        voidDocumentConfirm: res.data.voidOriginal
      }});
      return res.data.newDocumentNumber;
    });
  } catch (e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch({ type: UNAUTHORISED_COPY_DOCUMENT }); return;
    } else {
      dispatch(apiCallFailed(e.response));
    }
  }
};

export const checkCopyCertificate = (documentNumber) => async (dispatch, _getState, {orchestrationApi}) => {
  const config = {
    headers: {
      documentnumber: documentNumber
    }
  };

  try {
    const res = await orchestrationApi.get('/check-copy-certificate', config);
    const { canCopy } = res.data;
    if (!canCopy) {
      dispatch(unauthorised());
    }
  } catch(e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch(unauthorised());
      throw new Error(`Unauthorised access ${e.response.status}`);
    } else {
      dispatch(apiCallFailed(e.response));
      throw new Error(`An error has occurred ${e.response.status} (${e.response.statusText})`);
    }
  }

  function unauthorised() { return { type: UNAUTHORISED_COPY_DOCUMENT }; }
};

export const addCopyDocument = (data) => (dispatch) => dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: data });
export const clearErrors = () => (dispatch) =>  dispatch({ type: CLEAR_ERRORS });
export const clearCopyDocument = () => (dispatch) => dispatch({ type: CLEAR_COPY_DOCUMENT });
export const unauthorisedCopyDocument = () => (dispatch) => dispatch({ type: UNAUTHORISED_COPY_DOCUMENT });