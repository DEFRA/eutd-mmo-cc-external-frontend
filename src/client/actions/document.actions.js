// actions
import {isError, showFullPageError} from './index';

export const GOT_ALL_DOCUMENTS          = 'GOT_ALL_DOCUMENTS';
export const GET_ALL_DOCUMENTS_FAILED   = 'GET_ALL_DOCUMENTS_FAILED';
export const GET_COMPLETED_DOCUMENT = 'GET_COMPLETED_DOCUMENT';
export const GET_PENDING_DOCUMENT = 'GET_PENDING_DOCUMENT';


// action creators
export const getAllDocuments = (type, month, year) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/documents/${year}/${month}?type=${type}`);
    return dispatch({ type: GOT_ALL_DOCUMENTS, payload: res.data });

  } catch(e) {
    if( isError(e) ) dispatch(showFullPageError());
    //throw new Error(e.response);
  }
};

export const getCompletedDocument = (documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/document/${documentNumber}`);
    if (res.data && res.data.documentStatus === 'COMPLETE') {
      return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: res.data });
    }
    return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: null });
  }
  catch (e) {
    return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: null });
  }
};

export const getPendingDocument = (documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/document/${documentNumber}`);
    if (res.data && res.data.documentStatus === 'PENDING') {
      return dispatch({ type: GET_PENDING_DOCUMENT, payload: res.data });
    }
    return dispatch({ type: GET_PENDING_DOCUMENT, payload: null });
  }
  catch (e) {
    return dispatch({ type: GET_PENDING_DOCUMENT, payload: null });
  }
};