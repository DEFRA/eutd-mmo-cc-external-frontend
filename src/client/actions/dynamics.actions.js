const {parseContactsResponse, parseAddressResponse, parseAccountsResponse} = require('../../helpers/dynamix');

export const dynamicsActionTypes = {
  USER_DETAILS_REQUESTED: 'user-details/requested',
  USER_DETAILS_LOADED: 'user-details/loaded',
  USER_DETAILS_FAILED: 'user-details/failed',
  ADDRESS_DETAILS_REQUESTED: 'address-details/requested',
  ADDRESS_DETAILS_LOADED: 'address-details/loaded',
  ADDRESS_DETAILS_FAILED: 'address-details/failed',
  ACCOUNT_DETAILS_REQUESTED: 'account-details/requested',
  ACCOUNT_DETAILS_LOADED: 'account-details/loaded',
  ACCOUNT_DETAILS_FAILED: 'account-details/failed',
  EXPORTER_NAME_PRE_LOADED: 'export-certificate/exporter-name-preloaded',
  EXPORTER_ADDRESS_PRE_LOADED: 'export-certificate/exporter-address-preloaded',
  EXPORTER_ADDRESS_PRE_LOAD_FAILED: 'export-certificate/exporter-address-preload-failed',
  EXPORTER_COMPANY_NAME_PRE_LOADED: 'export-certificate/exporter-company-name-preloaded'
};


export const fetchUserDetailsFromDynamics = () => async (dispatch, getState, { dynamixApi } ) => {
  const url = '/user-details';
  try {
    const res = await dynamixApi.get(url);
    let userdetails = parseContactsResponse(res);
    dispatch(success(userdetails));
    if (userdetails && userdetails.length > 0) {
      dispatch(exporterNamePreloaded(userdetails[0].firstName, userdetails[0].lastName, userdetails[0].contactId));
    }
  } catch(error) {
    dispatch(failed('Failed server call to get user details'));
  }

  function success(userdetails) { return { type: dynamicsActionTypes.USER_DETAILS_LOADED, userdetails }; }
  function failed(error) { return { type: dynamicsActionTypes.USER_DETAILS_FAILED, error }; }
  function exporterNamePreloaded(firstName, lastName, contactId) { return { type: dynamicsActionTypes.EXPORTER_NAME_PRE_LOADED, firstName, lastName, contactId }; }
};

export const fetchAccountDetailsFromDynamics = () => async (dispatch, getState, { dynamixApi } ) => {
  function success(accounts) { return { type: dynamicsActionTypes.ACCOUNT_DETAILS_LOADED, accounts }; }
  function failed(error) { return { type: dynamicsActionTypes.ACCOUNT_DETAILS_FAILED, error }; }
  function exporterCompanyNamePreloaded(name, accountId) { return { type: dynamicsActionTypes.EXPORTER_COMPANY_NAME_PRE_LOADED, name, accountId }; }
  const url = '/accounts';

  try {
    const res = await dynamixApi.get(url);
    let accounts = parseAccountsResponse(res);
    dispatch(success(accounts));
    if (accounts && accounts.length > 0) {
      dispatch(exporterCompanyNamePreloaded(accounts[0].name, accounts[0].accountId));
    }
  } catch(error) {
    dispatch(failed('Failed server call to get account details'));
  }
};

export const fetchAddressDetailsFromDynamics = () => (dispatch, getState, { dynamixApi } ) => {
  const url = '/addresses';
  return dynamixApi.get(url)
    .then(
      res => {
        let addresses = parseAddressResponse(res);
        dispatch(success(addresses));
        if (addresses && addresses.length > 0) {
          dispatch(exporterAddressPreloaded(addresses[0]));
        }
      }
    )
    .catch((error) => {
      console.error(error);
      dispatch(failed('Failed server call to get address details'));
      dispatch(exporterAddressPreloadFailed());
    });

  function success(addresses) { return { type: dynamicsActionTypes.ADDRESS_DETAILS_LOADED, addresses }; }
  function failed(error) { return { type: dynamicsActionTypes.ADDRESS_DETAILS_FAILED, error }; }
  function exporterAddressPreloaded(address) { return { type: dynamicsActionTypes.EXPORTER_ADDRESS_PRE_LOADED, address }; }
  function exporterAddressPreloadFailed() { return { type: dynamicsActionTypes.EXPORTER_ADDRESS_PRE_LOAD_FAILED }; }
};

export const getAddressArrayFromDynamicsResponse = function(dynamicsResponse) {

  let addressArray = [];
  let addrOne = '';

  if (dynamicsResponse.address.premises && dynamicsResponse.address.premises !== null && dynamicsResponse.address.premises.length > 0) {
    addrOne += dynamicsResponse.address.premises;
    addrOne += ', ';
  }

  if (dynamicsResponse.address.subbuildingname && dynamicsResponse.address.subbuildingname !== null && dynamicsResponse.address.subbuildingname.length > 0) {
    addrOne += dynamicsResponse.address.subbuildingname;
    addrOne += ', ';
  }

  if (dynamicsResponse.address.buildingname && dynamicsResponse.address.buildingname !== null && dynamicsResponse.address.buildingname.length > 0) {
    addrOne += dynamicsResponse.address.buildingname;
    addrOne += ', ';
  }

  if (dynamicsResponse.address.street && dynamicsResponse.address.street !== null && dynamicsResponse.address.street.length > 0) {
    var partsOfStr = dynamicsResponse.address.street.split(',');
    if (partsOfStr.length > 1) {
      addrOne += partsOfStr[0].trim();
      addressArray.push(addrOne);
      addressArray.push(dynamicsResponse.address.street.slice(partsOfStr[0].length + 1).trim());
    } else {
      addrOne += dynamicsResponse.address.street;
      addressArray.push(addrOne);
    }
  } else {
    addressArray.push(addrOne);
  }

  if (dynamicsResponse.address.locality && dynamicsResponse.address.locality != null && dynamicsResponse.address.locality.length > 0) {
    addressArray.push(dynamicsResponse.address.locality);
  }
  if (dynamicsResponse.address.dependantLocality && dynamicsResponse.address.dependantLocality != null && dynamicsResponse.address.dependantLocality.length > 0) {
    addressArray.push(dynamicsResponse.address.dependantLocality);
  }
  if (dynamicsResponse.address.towntext && dynamicsResponse.address.towntext != null && dynamicsResponse.address.towntext.length > 0) {
    addressArray.push(dynamicsResponse.address.towntext);
  }
  if (dynamicsResponse.address.county && dynamicsResponse.address.county != null && dynamicsResponse.address.county.length > 0) {
    addressArray.push(dynamicsResponse.address.county);
  }

  return addressArray;
};

export const getAddressModelFromDynamicsResponse = function(dynamicsResponse) {

  let addressArray = getAddressArrayFromDynamicsResponse(dynamicsResponse);
  let model = {};
  model.addressOne = addressArray[0];

  if (dynamicsResponse.address.postcode && dynamicsResponse.address.postcode != null && dynamicsResponse.address.postcode.length > 0) {
    model.postcode = dynamicsResponse.address.postcode;
  } else if (dynamicsResponse.address.internationalpostalcode && dynamicsResponse.address.internationalpostalcode != null
    && dynamicsResponse.address.internationalpostalcode.length > 0) {
    model.postcode = dynamicsResponse.address.internationalpostalcode;
  }

  model._dynamicsAddress = dynamicsResponse.address._dynamicsAddress;

  let arrLength = addressArray.length;
  if (arrLength > 1) {
    model.townCity = addressArray[arrLength - 1];
    arrLength = arrLength - 1;
  }

  let addrTwo = '';
  if (arrLength > 1) {
    for (let idx = 1; idx < arrLength; idx++) {
      addrTwo += addressArray[idx];
      if (idx < arrLength - 1) {
        addrTwo += ', ';
      }
    }
  }
  model.addressTwo = addrTwo;
  model.buildingNumber = dynamicsResponse.address.premises;
  model.subBuildingName = dynamicsResponse.address.subbuildingname;
  model.buildingName = dynamicsResponse.address.buildingname;
  model.streetName = dynamicsResponse.address.street;
  model.townCity = dynamicsResponse.address.towntext || dynamicsResponse.address.locality;
  model.county = dynamicsResponse.address.county;
  model.country = dynamicsResponse.address.country;

  return model;
};

export const dynamicsActions = {
  fetchUserDetailsFromDynamics,
  fetchAccountDetailsFromDynamics,
  fetchAddressDetailsFromDynamics
};

