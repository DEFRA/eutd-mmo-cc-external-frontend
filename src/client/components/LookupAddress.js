import React from 'react';
import { Link } from 'react-router-dom';
import _ from 'lodash';
import { useTranslation } from 'react-i18next';

 const LookupAddress = ({
  addressOne,
  townCity,
  postcode,
  addressType,
  changeAddressLink,
  changeAddressHandler,
  unsavedFacilityName
}) => {
  const {t} = useTranslation();
  return (
  _.isEmpty(addressOne) && _.isEmpty(postcode) ? (
    <>
      <p id="lookup-address">{t('sdAddStorageFacilityDetailsAddressText', {addressType: addressType})}</p>
      <p id="address-link-wrapper">
        <Link
          id="add-lookup-address-link"
          className="change-address-link"
          to={changeAddressLink}
           onClick={() => changeAddressHandler({unsavedFacilityName}) }
        >
          {t('sdAddStorageFacilityDetailsAddAddressText', {addressType: addressType})}
        </Link>
      </p>
    </>
  ) : (
    <>
        <p id="lookup-address">
        {addressOne}
        <br />
        {townCity}
        <br />
        {postcode}
      </p>
      <p id="address-link-wrapper">
        <Link
            id="change-lookup-address-link"
          className="change-address-link"
          to={changeAddressLink}
         onClick={() => changeAddressHandler() }
        >
          {t('commonWhatExportersAddressChangeLink')}
          <span className="govuk-visually-hidden">{t('sdFacilityDetailsVHiddenChangeText')} {t('sdFacilityDetailsVHiddenAddressText')}</span>
        </Link>
      </p>
    </>
  )
 )  
}

export default LookupAddress