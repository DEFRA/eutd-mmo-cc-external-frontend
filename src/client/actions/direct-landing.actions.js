import { beginApiCall, apiCallFailed, isForbidden, showFullPageError, CLEAR_ERRORS, isSilverLineError } from './index';
import { exportPayloadActionTypes } from '../actions';


export const GET_DIRECT_LANDING_PRODUCTS = 'get_direct_landing_products';
export const CLEAR_DIRECT_LANDING = 'clear_direct_landing';
export const DIRECT_LANDING_PRODUCTS_FAILED = 'direct_landing_products_failed';
export const POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'post_direct_landing_products_unauthorised';
export const GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'get_direct_landing_products_unauthorised';
export const POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR = 'post_direct_landing_prodcts_waf_error';

export const getDirectLandings =
  (documentNumber) =>
  async (dispatch, _getState, { orchestrationApi }) => {
    const config = { headers: { documentnumber: documentNumber } };

    try {
      dispatch(beginApiCall());

      const res = await orchestrationApi.get(
        '/export-certificates/export-payload/direct-landings',
        config
      );

      await dispatch({ type: GET_DIRECT_LANDING_PRODUCTS, payload: res.data });
    } catch (e) {
      if (isForbidden(e)) {
        dispatch({ type: GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
      } else {
        dispatch(apiCallFailed(e.response));
        throw new Error(
          `An error has occurred ${e.response.status} (${e.response.statusText})`
        );
      }
    }
  };

export const clearDirectLanding = () => ({ type: CLEAR_DIRECT_LANDING });

export const upsertDirectLanding = (landing, documentNumber = undefined) => (dispatch, getState, { orchestrationApi } ) => {
  const config = (documentNumber)
    ? {
        headers: {
          documentnumber: documentNumber
        }
    }
    : {};

  return orchestrationApi.post('/export-certificates/direct-landing/validate', landing, config)
    .then(
      res => {
        dispatch(landingUpserted(res.data));
        dispatch(clearErrors());
      },
      error => {
        if (error.response.status === 400) {
          dispatch(landingUpserted(error.response.data));
          throw new Error();
        } else if (isSilverLineError(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR, supportID: error.response.data.match(/\d+/)[0] });
        } else if (isForbidden(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
        } else {
          dispatch(showFullPageError());
        }
      });

  function landingUpserted(exportPayload) { return { type: exportPayloadActionTypes.LANDING_UPSERTED, exportPayload, errors: exportPayload.errors }; }
  function clearErrors() { return { type: CLEAR_ERRORS }; }
};
