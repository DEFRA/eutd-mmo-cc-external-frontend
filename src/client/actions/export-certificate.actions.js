import { showSummaryCertificateErrors } from './certificate.actions';

export const exportCertificateActionTypes = {
  EXPORT_CERT_CREATE_REQUEST: 'export-certificate/create',
  EXPORT_CERT_CREATE_REQUEST_SUCCESS: 'export-certificate/create/success',
  EXPORT_CERT_CREATE_REQUEST_FAILED: 'export-certificate/create/failed'
};

export const createExportCertificate = (currentUri, journey, noOfVessels, documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? {headers: {documentNumber: documentNumber}}
    : {};

  const { data } = await orchestrationApi.get('/client-ip');

  return await orchestrationApi.post('/export-certificates/create', { journey, data },config)
    .then(
      res => {
        let certDetails = res.data;
        if (certDetails.status && certDetails.status === 'invalid catch certificate') {
          return 'ERROR';
        }
        window.dataLayer = window.dataLayer || [];
        window.dataLayer.push({
          'event': 'completedCatchCertificate',
          'numberOfVessels': noOfVessels
        });
        return dispatch(exportCertificateCreated(certDetails));
      },
      error => {
        dispatch(exportCertificateCreateFailed(error.toString()));
        if(error.response && error.response.data) {
          return dispatch(showSummaryCertificateErrors(error.response.data));
        }
      })
    .catch((err) => {
      dispatch(exportCertificateCreateFailed(err));
      throw(err);
    });

  function exportCertificateCreated(certDetails) { return { type: exportCertificateActionTypes.EXPORT_CERT_CREATE_REQUEST_SUCCESS,  payload: { data: certDetails } }; }
  function exportCertificateCreateFailed(errorRes) { return { type: exportCertificateActionTypes.EXPORT_CERT_CREATE_REQUEST_FAILED, errorRes }; }
};

export const getExportCertificateFromParams = (params) => (dispatch) => {
  return dispatch({type: exportCertificateActionTypes.EXPORT_CERT_CREATE_REQUEST_SUCCESS, payload: {data: params }});
};