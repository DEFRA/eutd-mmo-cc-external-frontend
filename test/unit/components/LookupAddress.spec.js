import React from 'react';
import { mount } from 'enzyme';
import  LookupAddress  from '../../../src/client/components/LookupAddress';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Link } from 'react-router-dom';
import thunk from 'redux-thunk';
import {changeStorageFacilityAddress} from '../../../src/client/actions/index';
import { render } from '@testing-library/react';

const storeData = {
    addAnotherProduct: 'notset',
    addAnotherStorageFacility: 'notset',
    catches: [{
      product: 'Asda squid burgers',
      commodityCode: '123456',
      certificateNumber: '12345',
      productWeight: '1000',
      dateOfUnloading: '25/01/2019',
      placeOfUnloading: 'Dover',
      transportUnloadedFrom: 'Car',
      weightOnCC: '500'
    }],
    storageFacilities: [
      {
        facilityName: 'FACILITY 1',
        facilityAddressOne: 'BUILDING 1, STREET 1',
        facilityBuildingName: 'BUILDING 1',
        facilityStreetName: 'STREET 1',
        facilityTownCity: 'TOWN 1',
        facilityCounty: 'COUNTY 1',
        facilityCountry: 'COUNTRY 1',
        facilityPostcode: 'POSTCODE 1'
      }
    ]
};

const expectedActions = [
  { type: 'change_storage_facility_address', payload: {unsavedFacilityName: 'Facilitity 1'}}
];


describe('Lookup Address Component if address is present', () => {
  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res({});
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  const store = mockStore({storeData});
 
  const props = {
    addressOne: 'LANCASTER HOUSE, HAMPSHIRE COURT',
    townCity: 'NEWCASTLE UPON TYNE',
    postcode: 'E11JR',
    addressType: 'storage facility',
    changeAddressLink:
      '/create-storage-document/GBR-2021-SD-67C7DC640/what-storage-facility-address/1',
    changeAddressHandler: jest.fn(),
  };
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <LookupAddress {...props} />
      </MemoryRouter>
    </Provider>
  );
  
  
  it('Should render the Lookup Address component', () => {
    expect(wrapper).toBeDefined();
  });

  it('Should render message "An address must be added for this storage facility."', () => {
    expect(wrapper.find('#lookup-address').exists()).toBeTruthy();
    expect(wrapper.find('#lookup-address').type()).toBe('p');
    expect(wrapper.find('#lookup-address').text()).toBe('LANCASTER HOUSE, HAMPSHIRE COURTNEWCASTLE UPON TYNEE11JR');
  });

  it('Should render change address link', () => {
    expect(wrapper.find('#change-lookup-address-link').first().exists()).toBeTruthy();
    expect(wrapper.find('#change-lookup-address-link').first().type()).toBe(Link);
    expect(wrapper.find('#change-lookup-address-link').first().text()).toBe('Changestorage facility address');
  });
  it('will handle clicking to add an address', () => {
    wrapper.find('#address-link-wrapper').find('Link').simulate('click');
    store.dispatch(changeStorageFacilityAddress({unsavedFacilityName: 'Facilitity 1'}));
    expect(store.getActions()).toEqual(expectedActions);
  });
  it('should take a snapshot of the whole page when address', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
});


describe('Address Lookup Component if address is not present', () => {
  const mockStore = configureStore([
    thunk.withExtraArgument({
      orchestrationApi: {
        get: () => {
          return new Promise((res) => {
            res({});
          });
        },
        post: () => {
          return new Promise((res) => {
            res({});
          });
        },
      },
    }),
  ]);

  const store = mockStore({storeData});

  const props = {
    addressOne: '',
    townCity: '',
    postcode: '',
    addressType: 'storage facility',
    changeAddressLink:
      '/create-storage-document/GBR-2021-SD-67C7DC640/what-storage-facility-address/1',
    changeAddressHandler: jest.fn(),
  };
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <LookupAddress {...props} />
      </MemoryRouter>
    </Provider>
  );
  

  it('Should render the Lookup Address component', () => {
    expect(wrapper).toBeDefined();
  });

  it('Should render link to add facility address', () => {
    expect(wrapper.find('#add-lookup-address-link').exists()).toBeTruthy();
    expect(wrapper.find('#add-lookup-address-link').first().type()).toBe(Link);
    expect(wrapper.find('#add-lookup-address-link').first().text()).toBe('Add the storage facility address');
  });

  it('will handle clicking to add an address', () => {
      wrapper.find('#address-link-wrapper').find('Link').simulate('click');
      store.dispatch(changeStorageFacilityAddress({unsavedFacilityName: 'Facilitity 1'}));
      expect(store.getActions()).toEqual(expectedActions);
  });
    
  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });
  
});

