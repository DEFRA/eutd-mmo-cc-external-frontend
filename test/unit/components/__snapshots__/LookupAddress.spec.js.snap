// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address Lookup Component if address is not present should take a snapshot of the whole page 1`] = `
<div>
  <p
    id="lookup-address"
  >
    An address must be added for this storage facility.
  </p>
  <p
    id="address-link-wrapper"
  >
    <a
      class="change-address-link"
      href="/create-storage-document/GBR-2021-SD-67C7DC640/what-storage-facility-address/1"
      id="add-lookup-address-link"
    >
      Add the storage facility address
    </a>
  </p>
</div>
`;

exports[`Lookup Address Component if address is present should take a snapshot of the whole page when address 1`] = `
<div>
  <p
    id="lookup-address"
  >
    LANCASTER HOUSE, HAMPSHIRE COURT
    <br />
    NEWCASTLE UPON TYNE
    <br />
    E11JR
  </p>
  <p
    id="address-link-wrapper"
  >
    <a
      class="change-address-link"
      href="/create-storage-document/GBR-2021-SD-67C7DC640/what-storage-facility-address/1"
      id="change-lookup-address-link"
    >
      Change
      <span
        class="govuk-visually-hidden"
      >
        storage facility
         
        address
      </span>
    </a>
  </p>
</div>
`;
